core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -256,15 +269,29 @@ class Partition(val topicPartition: TopicPartition,
     tags
   )
 
-  def isUnderReplicated: Boolean =
-    isLeader && inSyncReplicaIds.size < allReplicaIds.size
+  def isUnderReplicated: Boolean = assignmentState match {
+    case OngoingReassignmentState(_, _, originalReplicas, _, isr) => isLeader && (originalReplicas.size - isr.size) > 0
```

Comment:
There are a few cases like this where we could pull the logic into `AssignmentState`. For example:
```scala
trait AssignmentState {
  def replicationFactor: Int
}

def isUnderReplicated: Boolean = isLeader && (assignmentState.replicationFactor - isr.size) > 0
```

Reply:
Thanks, it makes the code easier to read

Reply:
One thing though: do we consider the replication factor to be the original as long as the replication itself isn't finished?
Specifically:
```
trait AssignmentState {
  def replicas: Seq[Int]
  def replicationFactor: Int = replicas.size
}

case class OngoingReassignmentState(addingReplicas: Seq[Int],
                                    removingReplicas: Seq[Int],
                                    originalReplicas: Seq[Int],
                                    replicas: Seq[Int]) extends AssignmentState {

  override def replicationFactor = originalReplicas.size
}

case class SimpleAssignmentState(replicas: Seq[Int]) extends AssignmentState
```

Reply:
Yes, what I documented in the KIP is this: 
```
isUnderReplicated == size(original assigned replicas) - size(isr) > 0
```